NavigationToolbar2Tk: make packing optional.

jmehne opened the pull request:
## PR Summary

Currently, `NavigationToolbar2Tk` cannot be easily used with the grid layout manager.
When initalizing the toolbar, it is immediately [packed](https://github.com/matplotlib/matplotlib/blob/70c8a4fad05a9fd964fc4db283b97940246f302b/lib/matplotlib/backends/_backend_tk.py#L585): `self.pack(side=tk.BOTTOM, fill=tk.X)`

I think this is unfortunate behavior, because:

- It is unexpected.
Usually, you initialize a widget and afterwards it must be added to one of the layout managers.
- It makes it hard to use the toolbar with the grid layout manager.
You need to call `pack_forget()` on the toolbar first, before you can add it to the grid layout manager.
- When resizing a window that contains the canvas and the toolbar in the same frame: how do you make sure, that the toolbar is *not* resized first and thus disappears before the canvas widget?
  You need to initialize it *before* the canvas, which again, is not super obvious.
  But then, how do you place the toolbar on top of the canvas - it is packed with `side=tk.BOTTOM`, so basically you have to call `pack_forget` and repack it with `side=tk.TOP`.
  Or, you initialize the canvas before the toolbar, but then you get the resizing problem again.

I assume we cannot change the default behavior for backwards compatibility, so this PR makes packing optional.

No idea how to write a test for this. Couldn't get the tests to run locally on master, so I'm relying on the CI tests.
## PR Checklist

- [ ] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

ImportanceOfBeingErnest commented on lib/matplotlib/backends/_backend_tk.py:
```diff
@@ -507,16 +507,23 @@ class NavigationToolbar2Tk(NavigationToolbar2, tk.Frame):
+        If True, add the toolbar to the parent's pack manager's packing list
+        during initialization with *side='bottom'* and *fill='x'*.
+        If you want to use the toolbar with a different layout manager, you
+        must first call *pack_forget* on the toolbar object.
```
I was confused by this sentence when first reading it, After all, the change proposed here is meant to prevent the need to call `pack_forget`. Maybe move it to the bottom and let it say

> If you want to use the toolbar with a different layout manager, use ``pack_toolbar=False``.

jmehne commented:
I changed the sentence accordingly.

I also changed the "Embedding in Tk" example again to use the new parameter, because I believe it is usually the better option:

- If you want to use pack, it is now easier to infer how to place the toolbar at the top.
- If you don't want to use pack, it should just work.
- Explicit is better than implicit :)

timhoffm commented on lib/matplotlib/backends/_backend_tk.py:
```diff
@@ -507,16 +507,22 @@ class NavigationToolbar2Tk(NavigationToolbar2, tk.Frame):
-        the tk.Window which owns this toolbar
-
+        The tk.Window which owns this toolbar.
+    pack_toolbar : bool, optional
```
```suggestion
    pack_toolbar : bool, default: True
```
We're now using this shortend version, when possible.

timhoffm commented on lib/matplotlib/backends/_backend_tk.py:
```diff
@@ -507,16 +507,22 @@ class NavigationToolbar2Tk(NavigationToolbar2, tk.Frame):
+        during initialization with *side='bottom'* and *fill='x'*.
+        If you want to use the toolbar with a different layout manager, use
+        *pack_toolbar=False*.
+        Default: True
```
```suggestion
```

jmehne commented:
Thanks for the reviews! I changed the docstring accordingly.

jmehne commented:
Hm, please don't merge this yet. In the example the toolbar height gets suddenly resized. Will take a look later today.

anntzer commented on lib/matplotlib/backends/_backend_tk.py:
```diff
@@ -507,16 +507,21 @@ class NavigationToolbar2Tk(NavigationToolbar2, tk.Frame):
+        *pack_toolbar=False*.
     """
-    def __init__(self, canvas, window):
+    def __init__(self, canvas, window, pack_toolbar=True):
```
please make the argument keyword only

jmehne replied:
After some googling the signature is now:

`def __init__(self, canvas, window, *, pack_toolbar=True):`

I didn't know this was possible at all, so please correct me if there is a more canonical way to do it.

jmehne commented:
Okay, I fixed the example. I used `expand=True` for the toolbar in the example which didn't make sense.
Looks good now:

![tkinter_test](https://user-images.githubusercontent.com/7451790/65016833-83d06200-d925-11e9-8657-d590049c4023.gif)

anntzer commented on lib/matplotlib/backends/_backend_tk.py:
```diff
@@ -507,16 +507,21 @@ class NavigationToolbar2Tk(NavigationToolbar2, tk.Frame):
         # Avoid using self.window (prefer self.canvas.get_tk_widget().master),
         # so that Tool implementations can reuse the methods.
         self.window = window
+        self.pack_toolbar = pack_toolbar
```
This doesn't need to be a public attribute -- it's just there for coordination between `__init__` and `_init_toolbar` -- which gets called by `__init__` anyways.
In fact you could even move the call to `pack` out of `_init_toolbar` and put it here:
```
if pack_toolbar: self.pack(...)
```

jmehne replied:
I moved the call.

anntzer commented:
I agree that the old behavior looked strange (well, at least from a qt PoV; dunno what's standard with tk...); if you think we should ultimately switch the default to not packing, see e.g. the deprecation strategy used in https://github.com/matplotlib/matplotlib/pull/12380

jmehne commented:
I'm not really a Tkinter expert, so I don't feel like I can decide at the moment if the default behavior should ultimately be switched or not. My gut feeling says yes, but there is no urgency to switch so I think it's a good idea to just introduce the change and let it sit for a bit.

anntzer commented on lib/matplotlib/backends/_backend_tk.py:
```diff
@@ -507,17 +507,23 @@ class NavigationToolbar2Tk(NavigationToolbar2, tk.Frame):
+        The tk.Window which owns this toolbar.
+    pack_toolbar : bool, default: True
+        If True, add the toolbar to the parent's pack manager's packing list
+        during initialization with *side='bottom'* and *fill='x'*.
```
use double-backquotes rather than asterisks

jmehne commented:
Docstring updated to use double backticks.

timhoffm commented:
Thanks!
